DragaliaAPI/Features/Reward/RewardService.cs

```diff
@@ -0,0 +1,149 @@
+
+public class RewardService : IRewardService
+{
+    private readonly ILogger<RewardService> logger;
+    private readonly IInventoryRepository inventoryRepository;
+    private readonly IUserDataRepository userDataRepository;
+    private readonly IAbilityCrestRepository abilityCrestRepository;
+    private readonly IUnitRepository unitRepository;
+
+    private List<Entity> discardedEntities = new();
+    private List<Entity> presentEntites = new();
+    private List<Entity> presentLimitEntities = new();
+    private List<Entity> newEntities = new();
+    private List<ConvertedEntity> convertedEntities = new();
+
+    public RewardService(
+        ILogger<RewardService> logger,
+        IInventoryRepository inventoryRepository,
+        IUserDataRepository userDataRepository,
+        IAbilityCrestRepository abilityCrestRepository,
+        IUnitRepository unitRepository
+    )
+    {
+        this.logger = logger;
+        this.inventoryRepository = inventoryRepository;
+        this.userDataRepository = userDataRepository;
+        this.abilityCrestRepository = abilityCrestRepository;
+        this.unitRepository = unitRepository;
+    }
+
+    public async Task<GrantResult> GrantReward(Entity entity)
+    {
+        switch (entity.Type)
+        {
+            case EntityTypes.Chara:
+                return await this.RewardCharacter(entity);
+            case EntityTypes.Dragon:
+                await this.unitRepository.AddDragons((Dragons)entity.Id);
+                break;
+            case EntityTypes.Dew:
+                await this.userDataRepository.UpdateDewpoint(entity.Quantity);
+                break;
+            case EntityTypes.HustleHammer:
+                (await this.userDataRepository.UserData.SingleAsync()).BuildTimePoint +=
+                    entity.Quantity;
+                break;
+            case EntityTypes.Rupies:
+                await this.userDataRepository.UpdateCoin(entity.Quantity);
+                break;
+            case EntityTypes.Wyrmite:
+                await this.userDataRepository.GiveWyrmite(entity.Quantity);
+                break;
+            case EntityTypes.Wyrmprint:
+                return await this.RewardAbilityCrest(entity);
+            case EntityTypes.Material:
+                (
+                    await this.inventoryRepository.GetMaterial((Materials)entity.Id)
+                    ?? this.inventoryRepository.AddMaterial((Materials)entity.Id)
+                ).Quantity += entity.Quantity;
+                break;
+            default:
+                logger.LogWarning("Tried to reward unsupported entity {@entity}", entity);
+                return GrantResult.FailError;
+        }
+
+        this.newEntities.Add(entity);
+        return GrantResult.Added;
+    }
+
+    private async Task<GrantResult> RewardCharacter(Entity entity)
+    {
+        if (entity.Type is not EntityTypes.Chara)
+            throw new ArgumentException("Entity was not a character", nameof(entity));
+
```

Comment:
you could do `Charas chara = (Charas)entity.Id;` once to not have to cast multiple times